spec/requests/halls_spec.rb

```diff
@@ -0,0 +1,345 @@
+require 'rails_helper'
+
+RSpec.describe "/halls", type: :request do
+  let(:user) { create :user }
+  let(:manager) { create :user, email: "testmanager@test.com", role: :manager }
+  describe "GET /halls" do
+    context "when no user" do
+      it "redirects to sign_in" do
+        get("/halls")
+        expect(response).to redirect_to("/users/sign_in")
+      end
+
+      it "returns redirect status" do
+        get("/halls")
+        expect(response.status).to eq(302)
+      end
+    end
```

Comment:
You can extract the request from specs like so:
```suggestion
RSpec.describe "/halls", type: :request do
  let(:user) { create :user }
  let(:manager) { create :user, email: "testmanager@test.com", role: :manager }
  subject(:request) { get('/halls') }
  
  before { request }
  
  describe "GET /halls" do
    context "when no user" do
      it "redirects to sign_in" do
        expect(response).to redirect_to("/users/sign_in")
      end

      it "returns redirect status" do
        expect(response.status).to eq(302)
      end
    end
```
You can redefine request for each context as you need and it will allow you to remove a lot of repetition

Reply:
The problem here is that if the `before { request }` comes before describe block, or even context, the sign_in user helper does not work, so i defined subject and put `before { request }` inside context. It still does remove a lot of repetition.